### PR TITLE
Link static lib directly into .node, remove OSX hack.

### PR DESCRIPTION
**Pull request by eddyb**
Should allow building on all architectures supported by Node.js and Rust.
Tested that it builds and loads fine on Linux.

### CODE REVIEW

**Comment by diorahman** [comment]
Interesting, I got this from the compiled `hello.node` lib on linux 64:

``` js
> require('./build/Release/hello')
{ '\u0000\u0000\u0000\u0000\u0000': [Function] }
> for (var k in a) { a[k](); }
'hello node'
```

**Comment by eddyb** [comment]
@diorahman Yes, that's fixed by dherman/neon#22.

**Comment by diorahman** [comment]
@eddyb thanks!

**Comment by eddyb** [comment]
Turns out shared libs are not worth creating because `.node` is already a shared lib and it's easier to link everything into it statically.

**Comment by No9** [comment]
FYI when I follow the example on FreeBSD 11 CURRENT December node 4.2.3 i get 

```
node -e 'require("./")'
/xxx/hello/lib/index.js:3
console.log(neon.hello());

TypeError: neon.hello is not a function
    at Object.<anonymous> (/xxx/hello/lib/index.js:3:18)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at [eval]:1:1
    at Object.exports.runInThisContext (vm.js:54:17)
    at Object.<anonymous> ([eval]-wrapper:6:22)
```

But when I follow the steps by @diorahman in https://github.com/dherman/neon-cli/pull/6#issuecomment-167090983 I get the same positive outcome with no error :) 

BTW Loving this cool stuff <3

**Comment by eddyb** [comment]
@No9 It seems that @dherman did not update the crates.io version of neon - so you're not getting the fix at dherman/neon#22.

If you replace the crates.io version dependency on `neon` with a github one (https://github.com/dherman/neon should work), and run `cargo update` followed by `npm install`, you should get a fully functional `.node` addon.

**Comment by No9** [comment]
Thanks @eddyb for your help 

After trying that update I seem to have walked into an unrelated issue

```
/xxx/neon-cli/lib/create.js:64
    { type: 'input', name: 'author',      message: "author",           default: guess.author   },
TypeError: Cannot read property 'author' of undefined
    at wizard (/usr/home/anton/builds/neon-cli/lib/create.js:64:86)
    at Object.<anonymous> (/usr/home/anton/builds/neon-cli/bin/cli.js:64:3)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:136:18)
    at node.js:963:3
```

FWIW I am running neon-cli linked not installed I'll try and come back to it at some stage and work out why

**Comment by diorahman** [comment]
Yeah, updating the deps, in the e.g. hello's `Cargo.toml` and the ../neon is dherman/neon seems working:

```
[dependencies]
neon = { version = "0.1", path = "../neon" }
```

**Comment by eddyb** [comment]
@diorahman `git = "https://github.com/dherman/neon"` in `Cargo.toml` is what I meant (where you currently have `path = "../neon"`).

**Comment by diorahman** [comment]
OK.

Got this one:

```
/home/diorahman/Experiments/misc/rs/neon/src/hello/build/Release/hello.node
module.js:460
  return process.dlopen(module, path._makeLong(filename));
                 ^

Error: ../target/release/libhello.so: cannot open shared object file: No such file or directory
    at Error (native)
    at Object.Module._extensions..node (module.js:460:18)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.load (/home/diorahman/Experiments/misc/rs/neon/src/hello/node_modules/neon-bridge/lib/index.js:19:10)
    at Object.<anonymous> (/home/diorahman/Experiments/misc/rs/neon/src/hello/lib/index.js:1:97)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
```

Do you have any ideas? 

Previously, in order to check the error, I then copied the `target` directory to `..`. And it works. It seems there is a problem regarding to the process cwd.

**Comment by eddyb** [comment]
@diorahman You're not using the latest version of this PR, maybe? I've switched from linking a dylib into the `.node` to linking a staticlib, which seems to work much better for me.

File: templates/binding.gyp.hbs [hunk 1]
```diff
@@ -29,15 +29,14 @@ Required context:
...
-            "shared_lib": "target/{{build.cfg.cargo}}/<(SHARED_LIB_PREFIX){{project.name}}<(SHARED_LIB_SUFFIX)"
+            "static_lib": "target/{{build.cfg.cargo}}/<(SHARED_LIB_PREFIX){{project.name}}<(STATIC_LIB_SUFFIX)"
         },
 
         "sources": ["<(INTERMEDIATE_DIR)/binding.cc"],      {{~! built client-side }}
 
         "include_dirs": ["<!(neon include-path)"],          {{~! computed client-side }}
 
-        "libraries": ["../<(shared_lib)"],                  {{~! gyp weirdly determines libraries from a subdirectory so ../ is needed }}
+        "libraries": ["../<(static_lib)"],                  {{~! gyp weirdly determines libraries from a subdirectory so ../ is needed }}
```
Comment (diorahman) [inline templates/binding.gyp.hbs:39]: Here I guess we can use:

```
"libraries": ["<(PRODUCT_DIR)/../../<(static_lib)"],
```

File: templates/binding.gyp.hbs [hunk 2]
```diff
@@ -45,12 +44,6 @@ Required context:
             "outputs": ["../<(static_lib)"],
```
Comment (diorahman) [inline templates/binding.gyp.hbs:44]: The same case as the previous one. It avoids creating unnecessary `target` directory in `..`

```
"libraries": ["<(PRODUCT_DIR)/../../<(static_lib)"],
```

**Comment by diorahman** [comment]
@eddyb thanks! Now everything works as expected!

**Comment by dherman** [comment]
This is great. I'll follow up immediately with the one fix needed to make it work in OS X again.

**Pull request merged**